atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/TransactionPreCommitActions.java

```diff
@@ -18,81 +18,96 @@
+ */
 @ThreadSafe
 final class TransactionPreCommitActions {
 
-    static class PreCommitAction {
-        final Consumer<LongSupplier> action;
-        final int timestampCount;
+    static class PreCommitActionWrapper {
+        final PreCommitAction action;
+        final int numLeasedTimestamps;
 
-        PreCommitAction(Consumer<LongSupplier> action, int timestampCount) {
+        PreCommitActionWrapper(PreCommitAction action, int numLeasedTimestamps) {
             this.action = action;
-            this.timestampCount = timestampCount;
+            this.numLeasedTimestamps = numLeasedTimestamps;
         }
 
         @Override
         public boolean equals(Object obj) {
-            return obj instanceof PreCommitAction
-                    && timestampCount == ((PreCommitAction) obj).timestampCount
-                    && action.equals(((PreCommitAction) obj).action);
+            return obj instanceof PreCommitActionWrapper
+                    && numLeasedTimestamps == ((PreCommitActionWrapper) obj).numLeasedTimestamps
+                    && action.equals(((PreCommitActionWrapper) obj).action);
         }
 
         @Override
         public int hashCode() {
-            return action.hashCode();
+            int hash = 5381;
+            hash += (hash << 5) + action.hashCode();
+            hash += (hash << 5) + Integer.hashCode(numLeasedTimestamps);
+            return hash;
         }
     }
 
     static class PerLeaseActions {
-        final List<PreCommitAction> preCommitActions;
-        int timestampCount;
+        final List<PreCommitActionWrapper> preCommitActions;
+        int numLeasedTimestamps;
 
         PerLeaseActions() {
             preCommitActions = new ArrayList<>();
-            timestampCount = 0;
+            numLeasedTimestamps = 0;
         }
 
-        private PerLeaseActions(List<PreCommitAction> actions, int timestampCount) {
+        private PerLeaseActions(List<PreCommitActionWrapper> actions, int numLeasedTimestamps) {
             preCommitActions = actions;
-            this.timestampCount = timestampCount;
+            this.numLeasedTimestamps = numLeasedTimestamps;
         }
 
         PerLeaseActions copy() {
-            return new PerLeaseActions(new ArrayList<>(preCommitActions), timestampCount);
+            return new PerLeaseActions(new ArrayList<>(preCommitActions), numLeasedTimestamps);
         }
 
         @Override
         public boolean equals(Object obj) {
             return obj instanceof PerLeaseActions
-                    && timestampCount == ((PerLeaseActions) obj).timestampCount
+                    && numLeasedTimestamps == ((PerLeaseActions) obj).numLeasedTimestamps
                     && preCommitActions.equals(((PerLeaseActions) obj).preCommitActions);
         }
 
         @Override
         public int hashCode() {
-            return preCommitActions.hashCode();
+            int hash = 5381;
+            hash += (hash << 5) + preCommitActions.hashCode();
```

Comment:
This still has the thing where the hash can change at runtime if we add a new action right? Which is generally not ideal for hash based collections.

To write it more explicitly, what I was poking at when I asked this on the post-merge review was whether `PerLeaseActions` actually needs to have `equals`/`hashCode` that aren't tests of object equality (and if so you don't need to override it, the default java implementation would work?)